{% extends 'frontend/base.html' %}

{% block title %}Region{% endblock %}

{% block css %}
  <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/locuszoom.css"
        type="text/css"/>
{% endblock %}

{% block content %}
  <div class="container-fluid">
    <div class="row padtop">
      <div class="col-sm-10">
        <div id="home-searchbox">
          <form id="variant-search">
            <div class="input-group">
              <div class="input-group-prepend">
                  <a href="/" class="btn btn-secondary btn-sm" role="button" aria-pressed="true"><span class="fa fa-home"></span></a>
              </div>
              <input id="query-field" name="query-field" autocomplete="off" class="form-control"
                     type="text"
                     placeholder="Search for a variant, e.g. chr19:488506 or rs10424907" autofocus/>
              <div class="input-group-append">
                <button id="navigate-variant" class="btn btn-secondary" type="submit"><span
                    class="fa fa-search"></span></button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>

    <br>
    {#  FIXME: This label does not update as the plot is panned or zoomed. We can resolve this in the future when we move to a dynamic frontend #}
    <h1 style="margin-top: 1em;"><strong>Single-tissue eQTLs near <span id="header-symbol">{{ symbol }}</span> (chr{{ chrom }}:{{ "{:,}".format(start) }}-{{ "{:,}".format(end) }}) </strong></h1>
  </div>

  <div class="container-fluid">
    <div class="row">
      <div id="lz-plot">
      </div>
    </div>
  </div>

  <!-- Bootstrap dropdown menu to choose an additional tissue to plot -->
  <div class="container-fluid">
    <div class="row justify-content-start">
      <div class="col-sm">
        Set a default gene or tissue as the anchor<br>
        <div class="btn-group pr-1">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true" data-placement="top"
                title="Choose a new <b>anchor tissue</b> or <b>gene</b>. All other added plots will be based on these anchors: when you add a <b>new gene</b>, the eQTLs plotted will be between that gene and the <b>anchor tissue</b>; when you add a <b>new tissue</b>, the eQTLs plotted will be between that tissue and the <b>anchor gene</b>. <br>Changing either anchor will delete all other plots and generate a single new plot, with eQTLs for the anchor gene in the anchor tissue.">
            <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"><span class="fa fa-info-circle"></span><span class="fa fa-anchor"></span></button>
          </span>
          <div class="dropdown">
            <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdown_anchorgeneselect" data-toggle="dropdown" data-toggle-second="tooltip" data-placement="right"
                  aria-haspopup="true" aria-expanded="false" title="Choose a gene to serve as an anchor">
              <i><span class="anchorgene"></span></i>
            </button>
            <!-- Dynamically generate anchor gene list -->
            <div class="dropdown-menu scrollable-menu" aria-labelledby="dropdown_anchorgeneselect">
              {% for single_geneid, single_symbol in gene_list.items() %}
                <button class="dropdown-item anchorgeneselect-class" type="button" data-genesymbol="{{ single_symbol }}" data-geneid="{{ single_geneid }}"><i>{{ single_symbol }}</i></button>
              {% endfor %}
            </div>
          </div>

          <span class="d-inline-block">
            <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> in</button>
          </span>

          <div class="dropdown">
            <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdown_anchortissueselect" data-toggle="dropdown" data-toggle-second="tooltip" data-placement="right"
                aria-haspopup="true" aria-expanded="false" title="Choose a tissue to serve as an anchor">
                <span class="anchortissue"></span>
            </button>
            <!-- Dynamically generate list of anchor tissue buttons -->
            <div class="dropdown-menu scrollable-menu" aria-labelledby="dropdown_anchortissueselect">
                {% for single_tissue in tissue_list %}
                  <button class="dropdown-item anchortissueselect-class" type="button" value="{{ single_tissue }}">{{ single_tissue }}</button>
                {% endfor %}
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm">
        Add plot for a new gene in the anchor tissue<br>
        <div class="btn-group pr-1">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true"
                title="Add an eQTL plot for a gene in the current anchor tissue">
            <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> <span class="fa fa-plus"></span></button>
          </span>
          <div class="dropdown">
            <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdown_geneselect" data-toggle="dropdown" data-toggle-second="tooltip" data-placement="left"
                    aria-haspopup="true" aria-expanded="false" title="Choose a gene to add a new panel showing eQTLs between that gene and the current anchor tissue">
              Gene
            </button>
            <!-- Dynamically generate gene list -->
            <div class="dropdown-menu scrollable-menu" aria-labelledby="dropdown_geneselect">
              {% for single_geneid, single_symbol in gene_list.items() %}
               <button class="dropdown-item geneselect-class" type="button" data-genesymbol="{{ single_symbol }}" data-geneid="{{ single_geneid }}"><i>{{ single_symbol }}</i></button>
              {% endfor %}
            </div>
          </div>
          <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;">in</button>
          <button type="button" class="btn btn-outline-secondary anchor-tissue-label" style="pointer-events: none;">{{ tissue }}</button>
        </div>
      </div>
      <div class="col-sm">
        Add plot for the anchor gene in a new tissue
        <div class="btn-group pr-1">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true"
            title="Add a new tissue with the current anchor gene">
            <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> <span class="fa fa-plus"></span></button>
          </span>
          <button type="button" class="btn btn-outline-secondary anchor-gene-label" style="pointer-events: none;"><i>{{ symbol }}</i></button>
          <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;">in</button>
          <div class="dropdown">
            <button class="btn btn-secondary dropdown-toggle" type="button" id="dropdown_tissueselect" data-toggle="dropdown" data-toggle-second="tooltip" data-placement="right"
                  aria-haspopup="true" aria-expanded="false" title="Choose a tissue to add a new panel with eQTLs between that tissue and the current anchor gene">
            Tissue
            </button>
          <!-- Dynamically generate tissue buttons from tissue_list -->
            <div class="dropdown-menu scrollable-menu" aria-labelledby="dropdown_tissueselect">
              {% for single_tissue in tissue_list %}
              <button class="dropdown-item tissueselect-class" type="button" value="{{ single_tissue }}">{{ single_tissue }}</button>
              {% endfor %}
            </div>
          </div>
        </div>
      </div>
      <div class="col-sm">
        Change the plotted values
        <div class="btn-group pr-1">
          <form class="yaxis-display" id="transform-y">
            <div class="btn-group btn-group-toggle" data-toggle="buttons">
              <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-html="true"
                  title="Switches the y variable between -log<sub>10</sub>(P-value) and Normalized Effect Size (NES). Triangles indicate eQTLs for upregulation (pointing up) or downregulation (pointing down) of gene expression with P-values < 0.05.">
                <button type="button" class="btn btn-outline-secondary" style="pointer-events: none;"> <span class="fa fa-arrows-alt-v"></span> Y-Axis: </button>
              </span>
              <label class="btn btn-secondary active" data-toggle="tooltip" data-placement="top" data-html="true"
                  title="Display -log<sub>10</sub>(P-values) on the Y-axis">
                <input type="radio" name="y-options" autocomplete=off id="show-log-pvalue" value="log_pvalue"> P-value
              </label>
              <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top" data-html="true"
                  title="Displays Normalized Effect Sizes (NES) on the Y-axis. See <a href='https://www.gtexportal.org/home/documentationPage'>the GTEx Portal</a> for an explanation of NES.">
                <input type="radio" name="y-options" autocomplete=off id="show-beta" value="beta"> Effect Size
              </label>
              <label class="btn btn-secondary" data-toggle="tooltip" data-placement="top" data-html="true"
                  title="Displays <a href='https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1006646' target='_blank'>DAP-G</a> Posterior Inclusion Probabilities (PIP) on the Y-axis.<br>Cluster 1 denotes the cluster of variants (in LD with each other) with the strongest signal; cluster 2 denotes the set of variants with the next strongest signal; and so on.">
                <input type="radio" name="y-options" autocomplete=off id="show-pip" value="pip"> PIP
              </label>
            </div>
          </form>
        </div>
      </div>

    </div>
  </div>

<br>

<!-- 
  Since these links are based on Jinja template values, they currently do not update when we change anchor genes.
  We plan to update the mechanism handling URLs so that when you change the anchor gene, you are redirected to a new page.
  This will allow these links to update correctly when changing the anchor gene.
-->
<div class="container-fluid">
  <div class="card">
    <div class="row">
      <div class="col">
        <div class="card-body">
          <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top"
            title="External links for more information about this gene">
          <button class="btn btn-sm btn-secondary" style="pointer-events: none;"><span
            class="fa fa-secondary-circle"></span><span class="fa fa-info-circle"></span> More info about {{ symbol }} </button>
          </span>
          {% set short_gene_id = gene_id.split('.')[0] %}
          <a href="https://bravo.sph.umich.edu/freeze5/hg38/gene/{{ short_gene_id }}" target="_blank"
              class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
              data-placement="top" data-html=true
              title="Gene information from NHLBI's TOPMed program, containing 463 million variants observed in 62,784 individuals in data freeze 5. <b>Requires Google login</b>">
            BRAVO <span class="fa fa-external-link-alt"></span> </a>          
          <a href="https://gtexportal.org/home/gene/{{ symbol }}" target="_blank"
              class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
              data-placement="top" data-html=true
              title="Detailed information from the GTEx Portal, including both gene and exon expression, along with single-tissue eQTLs and sQTLs.">
            GTEx Portal <span class="fa fa-external-link-alt"></span> </a>
          <a href="https://gnomad.broadinstitute.org/gene/{{ symbol }}?dataset=gnomad_r3" target="_blank"
              class="btn btn-secondary btn-sm" role="button" aria-pressed="true" data-toggle="tooltip"
              data-placement="top"
              title="The Genome Aggregation Database (v3) at the Broad Institute, containing variant data from 71,702 sequenced genomes">
            gnomAD <span class="fa fa-external-link-alt"></span></a>
          <a href="http://pheweb.sph.umich.edu/gene/{{ symbol }}" target="_blank" class="btn btn-secondary btn-sm" role="button"
              aria-pressed="true" data-toggle="tooltip" data-placement="top"
              title="PheWeb summary of association results from 1,448 electronic health record-derived phenotypes tested against up to ~6,000 cases and ~18,000 controls with genotyped and imputed samples from the Michigan Genomics Initiative">
            MGI <span class="fa fa-external-link-alt"></span></a>
          <a href="http://pheweb.sph.umich.edu/SAIGE-UKB/gene/{{ symbol }}" target="_blank" class="btn btn-secondary btn-sm"
              role="button" aria-pressed="true" data-toggle="tooltip" data-placement="top"
              title="PheWeb summary of association results from the UK Biobank, with up to ~78k cases and ~409k controls, with binary outcomes analyzed with the SAIGE software">
            UKB-SAIGE <span class="fa fa-external-link-alt"></span></a>
          <a href="http://big.stats.ox.ac.uk/gene/{{ symbol }}" target="_blank" class="btn btn-secondary btn-sm" role="button"
              aria-pressed="true" data-toggle="tooltip" data-placement="top"
              title="Summary of 3,144 GWAS of Brain Imaging Derived Phenotypes (IDPs) in 9,707 participants from the UK Biobank, analyzed with the BGENIE software">
            UKB-Oxford BIG <span class="fa fa-external-link-alt"></span></a>
          <a href="http://www.ebi.ac.uk/gxa/search?geneQuery=[{'value':'{{symbol}}'}]" target="_blank" class="btn btn-secondary btn-sm" role="button"
              aria-pressed="true" dadta-toggle="tooltip" data-placement="top"
              title="The Expression Atlas is a project from the European Bioinformatics Institute (EMBL-EBI), with results from over 3,000 experiments from 40 different organisms, which have been manually reviewed, curated, and standardized.">
             Expression Atlas <span class="fa fa-external-link-alt"></span></a>
        </div>
      </div>
    </div>
  </div>
</div>

<div class="container-fluid">
  <div class="row">
    <div class="col">
      <span class="d-inline-block" tabindex="0" data-toggle="tooltip" data-placement="top" data-html="true"
          title="Created by Alan Kwong, Mukai Wang, Andy Boughton, Peter VandeHaar, and Hyun Min Kang. Source code can be found on <a href=https://github.com/statgen/pheget/>GitHub</a>.">
        <span class="badge badge-pill badge-secondary" style="pointer-events: none;">
        <span class="fa fa-lightbulb-o"></span> Credits
        </span>
      </span>
    </div>
  </div>
</div>

{% endblock %}

{% block javascript %}
  <script src="{{ url_for('static', filename='js/searchbox.js') }}" type="application/javascript"></script>

  {# These files must be included to make a LocusZoom plot #}
  <script src="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/locuszoom.vendor.min.js"
          type="application/javascript"></script>
  <script src="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/locuszoom.app.min.js"
          type="application/javascript"></script>

  <script type="application/javascript"
          src="https://cdn.jsdelivr.net/npm/locuszoom@{{ config.LOCUSZOOM_VERSION }}/dist/ext/lz-dynamic-urls.min.js"></script>

  <script src="{{ url_for('static', filename='js/common.js') }}" type="application/javascript"></script>
  <script src="{{ url_for('static', filename='js/region.js') }}" type="application/javascript"></script>

  <script type="application/javascript">
      'use strict';
      var ANCHOR_GENE = {{gene_id|tojson}};
      var ANCHOR_SYMBOL = {{symbol|tojson}};
      var ANCHOR_TISSUE = {{tissue|tojson}};
      var GENE_DICT = {{gene_list|tojson}};

      [window.plot, window.datasources] = makeSinglePlot(ANCHOR_GENE, ANCHOR_TISSUE, '#lz-plot', ANCHOR_SYMBOL);

      // Use omnisearch (https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch) to parse
      document.getElementById('variant-search').addEventListener('submit', function (event) {
          event.preventDefault();
          const variant = document.getElementById('query-field').value;
          if (!variant) {
              // TODO: If no search result was found, perhaps we can show error messaging instead of failing silently?
              return;
          }
          parseSearch(variant);
      });

      // Bootstrap dropdown selector for adding a new tissue
      var tissueElements = document.getElementsByClassName('tissueselect-class');
      Array.from(tissueElements).forEach((selectedTissue) => {
          selectedTissue.addEventListener('click', (event) => {
              addTrack(window.plot, window.datasources, ANCHOR_GENE, event.currentTarget.value, ANCHOR_SYMBOL);
          });
      });

      var geneElements = document.getElementsByClassName('geneselect-class');
      Array.from(geneElements).forEach((selectedGene) => {
          selectedGene.addEventListener('click', (event) => {
              addTrack(window.plot, window.datasources, event.currentTarget.dataset.geneid, ANCHOR_TISSUE, event.currentTarget.dataset.genesymbol);
          });
      });

      var anchortissueElements = document.getElementsByClassName('anchortissueselect-class');
      Array.from(anchortissueElements).forEach((selectedAnchorTissue) => {
          selectedAnchorTissue.addEventListener('click', (event) => {
            ANCHOR_TISSUE = event.currentTarget.value;
            document.querySelector('.anchortissue').innerHTML = ANCHOR_TISSUE;
            document.querySelector('.anchor-tissue-label').innerHTML = ANCHOR_TISSUE;
            let panel_ids = Array.from(plot.layout.panels, panel => panel.id);
            panel_ids.forEach(function(id){
              if (id !=='genes'){
                plot.removePanel(id);
              }
            });
            // FIXME: seems to be an ugly way of removing panels one by one
            // TODO: When changing anchors, simply redirect to new URL with the new anchors
            addTrack(window.plot, window.datasources, ANCHOR_GENE, ANCHOR_TISSUE, ANCHOR_SYMBOL);
          });
      });

      var anchorgeneElements = document.getElementsByClassName('anchorgeneselect-class');
      Array.from(anchorgeneElements).forEach((selectedAnchorGene) => {
          selectedAnchorGene.addEventListener('click', (event) => {
            ANCHOR_GENE = event.currentTarget.dataset.geneid;
            ANCHOR_SYMBOL = event.currentTarget.dataset.genesymbol;
            document.querySelector('.anchorgene').innerHTML = GENE_DICT[ANCHOR_GENE];
            document.querySelector('.anchor-gene-label').innerHTML = GENE_DICT[ANCHOR_GENE];
            document.getElementById('header-symbol').innerHTML = GENE_DICT[ANCHOR_GENE];
            let panel_ids = Array.from(plot.layout.panels, panel => panel.id);
            panel_ids.forEach(function(id){
              if (id !=='genes'){
                plot.removePanel(id);
              }
            });
            // FIXME: seems to be an ugly way of removing panels one by one
            // TODO: Same as above - redirect to new URL with new anchors
            addTrack(window.plot, window.datasources, ANCHOR_GENE, ANCHOR_TISSUE, ANCHOR_SYMBOL);
          });
      });

      document.forms['transform-y'].elements['y-options'].forEach(function (form) {
          form.onchange = function () {
              switchY_region(plot, this.value);
          };
      });

  // initialize labels for anchor tissue and gene
  document.querySelector('.anchortissue').innerHTML = ANCHOR_TISSUE;
  document.querySelector('.anchorgene').innerHTML = GENE_DICT[ANCHOR_GENE];

  </script>
{% endblock %}
